### DISCUSSION TITLE
Select from a read-only view

### DISCUSSION THREAD

**Discussion by rdong8** [sqlalchemy/sqlalchemy#12187] [Usage Questions] [answered]
I am trying to replicate this postgres query in sqlalchemy:

```sql
SELECT users, decrypted_secrets.decrypted_secret
FROM public.users
INNER JOIN vault.decrypted_secrets ON users.secret_id = decrypted_secrets.id;
```

Here, `public.users` is a table defined as a sqlalchemy model but `vault.decrypted_secrets` is a non-materialized view that is not. I don't want to recreate the view in sqlalchemy because I don't control it, and its definition might slightly change even though I'm only interested in 2 columns. Is there any way I can do this? The closest I've gotten seems to be this:

```python
class Base(AsyncAttrs, DeclarativeBase):
    pass


class User(Base):
    __tablename__ = "users"
    __table_args__ = {"schema": "public"}

    user_id: Mapped[int] = orm.mapped_column(BigInteger, primary_key=True)
    secret_id: Mapped[UUID] = orm.mapped_column(ForeignKey("vault.secrets.id"))


# This is a view that Supabase creates and manages
class DecryptedSecret(Base):
    __tablename__ = "decrypted_secrets"
    __table_args__ = {"schema": "vault"}
    __mapper_args__ = {
        "primary_key": ["id"],
    }

    id: Mapped[UUID] = orm.mapped_column(sa.UUID, primary_key=True)
    decrypted_secret: Mapped[str] = orm.mapped_column(String)


# ...setup code

statement = sa.select(User, DecryptedSecret.decrypted_secret).filter_by(user_id=0)

async with sessionmaker() as session:
    result = await session.scalars(statement)

# This gives me a User and not a (User, DecryptedSecret.decrypted_secret)
print(result.first())
```

**Comment by zzzeek** [comment, marked as answer]
>  I don't want to recreate the view in sqlalchemy because I don't control it, and its definition might slightly change even though I'm only interested in 2 columns. Is there any way I can do this?

absolutely, you can define any `Table` or mapped class with only the columns that you care about, there is no need to state other columns that you don't care about.

```py

statement = sa.select(User, DecryptedSecret.decrypted_secret).filter_by(user_id=0)

# This gives me a User and not a (User, DecryptedSecret.decrypted_secret)
print(result.first())
```

that query when you call upon `first()` will definitely give you a tuple of `(User, <something>)` since the SELECT is against those two different entities separately.  however in SQL you'll want to always JOIN those two entities together so they line up, somehting like

```
statement = sa.select(User, DecryptedSecret.decrypted_secret).join(User, DecryptedSecret).filter_by(user_id=0)
```

since you have a ForeignKey set up it should automatically figure out the criteria to join those two together.

**Comment by CaselIT** [comment]
```py
async with sessionmaker() as session:
    result = await session.scalars(statement)

# This gives me a User and not a (User, DecryptedSecret.decrypted_secret)
print(result.first())
```

scalars does return only the first column as a list of scalars. If you want a list of tuples use execute instead
